src/Microsoft.ML.Transforms/GcnTransform.cs

```diff
@@ -225,477 +255,725 @@ public void Save(ModelSaveContext ctx)
+        public static CommonOutputs.TransformOutput GcNormalize(IHostEnvironment env, LpNormalizingTransform.GcnArguments input)
+        {
+            var h = EntryPointUtils.CheckArgsAndCreateHost(env, "GcNormalize", input);
+            var xf = LpNormalizingTransform.Create(h, input, input.Data);
+            return new CommonOutputs.TransformOutput()
+            {
+                Model = new TransformModel(h, xf, input.Data),
+                OutputData = xf
+            };
         }
+    }
 
+    /// <summary>
+    /// Base estimator class for LpNorm and Gcn normalizers.
+    /// </summary>
+    public abstract class LpNormEstimatorBase : TrivialEstimator<LpNormalizingTransform>
+    {
         /// <summary>
-        /// Compute Standard Deviation. In case of both subMean and useStd are true, we technically need to compute variance
-        /// based on centered values (i.e. after subtracting the mean). But since the centered
-        /// values mean is approximately zero, we can use variance of non-centered values.
+        /// The kind of unit norm vectors are rescaled to. This enumeration is serialized.
         /// </summary>
-        private static Float StdDev(Float[] values, int count, int length)
-        {
-            Contracts.Assert(0 <= count && count <= length);
-            if (count == 0)
-                return 0;
-            // We need a mean to compute variance.
-            Float tmpMean = CpuMathUtils.Sum(values.AsSpan(0, count)) / length;
-            Float sumSq = 0;
-            if (count != length && tmpMean != 0)
-            {
-                // Sparse representation.
-                Float meanSq = tmpMean * tmpMean;
-                sumSq = (length - count) * meanSq;
-            }
-            sumSq += CpuMathUtils.SumSq(tmpMean, values.AsSpan(0, count));
-            return MathUtils.Sqrt(sumSq / length);
+        public enum NormalizerKind : byte
+        {
+            L2Norm = 0,
+            StdDev = 1,
+            L1Norm = 2,
+            LInf = 3
+        }
+
+        internal static class Defaults
+        {
+            public const NormalizerKind NormKind = NormalizerKind.L2Norm;
+            public const bool LpSubMean = false;
+            public const bool GcnSubMean = true;
+            public const bool UseStdDev = false;
+            public const float Scale = 1;
         }
 
         /// <summary>
-        /// Compute Standard Deviation.
-        /// We have two overloads of StdDev instead of one with <see cref="Nullable{Float}"/> mean for perf reasons.
+        /// Create a <see cref="LpNormEstimatorBase"/> that takes multiple pairs of columns.
         /// </summary>
-        private static Float StdDev(Float[] values, int count, int length, Float mean)
+        public LpNormEstimatorBase(IHostEnvironment env, params LpNormalizingTransform.ColumnInfoBase[] columns)
```

Comment:
>public [](start = 8, length = 6)

should be `protected

Reply:
I can't, this is base class, i can't make it less visible than it's children.

---
In reply to: [232437160](https://github.com/dotnet/machinelearning/pull/1582#discussion_r232437160) [](ancestors = 232437160)

Reply:
Surely you can

---
In reply to: [232767026](https://github.com/dotnet/machinelearning/pull/1582#discussion_r232767026) [](ancestors = 232767026,232437160)

Reply:
No I can't.

---
In reply to: [233619201](https://github.com/dotnet/machinelearning/pull/1582#discussion_r233619201) [](ancestors = 233619201,232767026,232437160)